### PR TITLE
gcc12, gcc13: always mangle __FILE__ into invalid store path

### PR DESCRIPTION
**Pull request by trofi**
Before the change macros like assert() embedded paths to source file into an executable binary. Examples are nlohmann_json embedding into nix:

    $ nix shell nixpkgs#nix nixpkgs#binutils-unwrapped nixpkgs#which
    $ strings $(which nix) | fgrep nlohmann_json
    /nix/store/5xih6daf5g3hpa0wc5vs2cgrhakn4s0j-nlohmann_json-3.11.2/include/nlohmann/json.hpp
    /nix/store/5xih6daf5g3hpa0wc5vs2cgrhakn4s0j-nlohmann_json-3.11.2/include/nlohmann/detail/output/serializer.hpp
    /nix/store/5xih6daf5g3hpa0wc5vs2cgrhakn4s0j-nlohmann_json-3.11.2/include/nlohmann/detail/conversions/to_chars.hpp
    /nix/store/5xih6daf5g3hpa0wc5vs2cgrhakn4s0j-nlohmann_json-3.11.2/include/nlohmann/detail/input/lexer.hpp
    /nix/store/5xih6daf5g3hpa0wc5vs2cgrhakn4s0j-nlohmann_json-3.11.2/include/nlohmann/detail/iterators/iter_impl.hpp
    /nix/store/5xih6daf5g3hpa0wc5vs2cgrhakn4s0j-nlohmann_json-3.11.2/include/nlohmann/detail/input/json_sax.hpp
    /nix/store/5xih6daf5g3hpa0wc5vs2cgrhakn4s0j-nlohmann_json-3.11.2/include/nlohmann/detail/iterators/iteration_proxy.hpp
    /nix/store/5xih6daf5g3hpa0wc5vs2cgrhakn4s0j-nlohmann_json-3.11.2/include/nlohmann/detail/input/parser.hpp

Sometimes these dependencies are not too large: nlohmann_json is "just" 1MB of headers code.

But sometimes the `-dev` outputs contain python scripts and other build-only dependencies. The example is `lttng-ust`.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by trofi** [comment]
And examples like `pipewire` reduced their closure down for about 20% (~90MB):

Before:

```
$ nix path-info -rsSh $(nix-build '<nixpkgs>' -A pipewire.out) | nl | tail -n1
   219  /nix/store/ff0w34nr807in3b1swmqklxy9g9v5hg9-pipewire-0.3.79 1.4M  543.4M
```

After:

```
$ nix path-info -rsSh $(nix-build ~/n -A pipewire.out) | nl | tail -n1
   207  /nix/store/hl4dffvc73nsh3zfbji0y7h9lcnrk14b-pipewire-0.3.79 1.3M  452.0M
```

**Comment by Mindavi** [comment]
This seems like a pretty safe change to me and will reduce closure size in (presumably) all sorts of places. My bash-foo is not up to speed, so I'll leave that to others.

**Comment by trofi** [comment]
Even dedupe is still able to fail `qemu` build, but a bit later:

```
-----------
Command line: `gcc -m64 -mcx16 /build/qemu-8.1.0/build/meson-private/tmpbyikv8nc/testfile.c -o /build/qemu-8.1.0/build/meson-private/tmpbyikv8nc/output.exe -D_FILE_OFFSET_BITS=64 -O0 -Wl,--start-group -laio -Wl,--end-group -Wl,--allow-shlib-undefined` -> 1
stderr:
gcc: fatal error: cannot execute '/nix/store/yabgsv2ahyxk7v8n1msp4z8ybaw12ham-gcc-14.0.0/libexec/gcc/x86_64-unknown-linux-gnu/14.0.0/cc1': execv: Argument list too long
compilation terminated.
-----------
```

Will try a response file instead. Converting to draft.

**Comment by trofi** [comment]
An attempt to use response files populated in setup hooks has at least 3 problems:
- `expandResponseParams` is empty in early bootstrap and that triggers a deficiency in `gcc-8`
- `setupHook` can't create files as `nix develop` executes `setupHooks` in a separate derivation
- something in response file file presence breaks `-lstdc++` file detection in `audiofile` package

**Comment by trofi** [comment]
Response files would not help shorter term either: https://trofi.github.io/posts/299-maximum-argument-count-on-linux-and-in-gcc.html

It might be simpler to patch `gcc` and `clang` directly until response file infrastructure is ready.

**Comment by trofi** [comment]
Filed https://gcc.gnu.org/PR111527 for longer term solution in `gcc`. Switched to direct `gcc` patching to mangle `NIX_STORE` references out of `__FILE__` macros.

The patches still are able to reduce `nix` and `pipewire` closure. Ready for review.

File: pkgs/development/compilers/gcc/patches/12/mangle-NIX_STORE-in-__FILE__.patch [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+Date: Fri, 22 Sep 2023 22:41:49 +0100
+Subject: [PATCH] gcc/file-prefix-map.cc: always mangle __FILE__ into invalid
+ store path
+
+Without the change `__FILE__` used in static inline functions in headers
+embed paths to header files into executable images. For local headers
+it's not a problem, but for headers in `/nix/store` this causes `-dev`
+inputs to be retained in runtime closure.
+
+Typical examples are `nix` -> `nlonhamm_json` and `pipewire` ->
```
Comment (Mindavi) [inline pkgs/development/compilers/gcc/patches/12/mangle-NIX_STORE-in-__FILE__.patch:12]: Typo: nlohmann

**Review by Mindavi — APPROVED** [review]
I think this is a fine solution and it probably won't hurt any users outside. __FILE__ is mostly an indication anyway.

File: pkgs/development/compilers/gcc/patches/12/mangle-NIX_STORE-in-__FILE__.patch (hunk 1 above)
Comment (trofi) [reply pkgs/development/compilers/gcc/patches/12/mangle-NIX_STORE-in-__FILE__.patch]: Oh, good catch! Fixed typo in 3 places.

**Comment by trofi** [comment]
As a longer-term workaround proposed limit increase in linux kernel as well: https://lkml.org/lkml/2023/9/24/381

File: pkgs/development/compilers/gcc/patches/12/mangle-NIX_STORE-in-__FILE__.patch [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    ...
+
+Mangled successfully.
+--- a/gcc/file-prefix-map.cc
++++ b/gcc/file-prefix-map.cc
+@@ -60,6 +60,9 @@ add_prefix_map (file_prefix_map *&maps, const char *arg, const char *opt)
+   maps = map;
+ }
+ 
++/* Forward declatration for a $NIX_STORE remap hack below. */
```
Comment (RaitoBezarius) [inline pkgs/development/compilers/gcc/patches/12/mangle-NIX_STORE-in-__FILE__.patch:40]:
```suggestion
+/* Forward declaration for a $NIX_STORE remap hack below. */
```

File: pkgs/development/compilers/gcc/patches/13/mangle-NIX_STORE-in-__FILE__.patch [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    ...
+
+Mangled successfully.
+--- a/gcc/file-prefix-map.cc
++++ b/gcc/file-prefix-map.cc
+@@ -69,6 +69,9 @@ add_prefix_map (file_prefix_map *&maps, const char *arg, const char *opt)
+   maps = map;
+ }
+ 
++/* Forward declatration for a $NIX_STORE remap hack below. */
```
Comment (RaitoBezarius) [inline pkgs/development/compilers/gcc/patches/13/mangle-NIX_STORE-in-__FILE__.patch:40]:
```suggestion
+/* Forward declaration for a $NIX_STORE remap hack below. */
```

**Review by RaitoBezarius — APPROVED** [review]
I do not have enough expertise in GCC to take responsibility. But the patch seems sane, the solution seems reasonable. I am in favor of this change.

**Comment by Artturin** [comment]
Building `pkgsCross.aarch64-multiplatform.sway`

**Comment by Ericson2314** [comment]
@trofi Have you tried to submit this patch upstream? I am recalling how I never got around to that with the Binutils patch, and then it became hard to maintain and that patch was puled out (by you no less :P).

I sorely miss that patch (now, because LLVM depends on libbfd for the LTO plugin, I always end up recompiling LLVM for new arches even though it is multitarget). I don't want you to lose this patch and miss it too. So I think the lesson is to submit upstream first.

To be clear I am definitely *not* saying we should block doing this on them accepting the patch. Rather that there should at least be an email thread we can link in Nixpkgs, and have the upstreaming process be started. Perfectly fine if it a takes a while for it to resolve itself.

File: pkgs/development/compilers/gcc/patches/12/mangle-NIX_STORE-in-__FILE__.patch (hunk 2 above)
Comment (trofi) [reply pkgs/development/compilers/gcc/patches/12/mangle-NIX_STORE-in-__FILE__.patch]: Updated the patch. Thank you!

File: pkgs/development/compilers/gcc/patches/13/mangle-NIX_STORE-in-__FILE__.patch (hunk 3 above)
Comment (trofi) [reply pkgs/development/compilers/gcc/patches/13/mangle-NIX_STORE-in-__FILE__.patch]: Updated the patch. Thank you!

**Comment by trofi** [comment]
Sent the problem statement with this proof-of-concept patch to `gcc` mailing list: https://gcc.gnu.org/pipermail/gcc/2023-October/242636.html

**Comment by Artturin** [comment]
`pkgsCross.aarch64-multiplatform.sway` built, no runtime closure size diff.

**Comment by trofi** [comment]
> `pkgsCross.aarch64-multiplatform.sway` built, no runtime closure size diff.

`sway` does not have a very big runtime closure. I found `pipewire` (and it's users) a bit different.

**Review by Artturin — APPROVED** [review]
Build-wise it works

**Comment by dotlambda** [comment]
This broke `python3.pkgs.deltachat`: https://github.com/NixOS/nixpkgs/pull/267870#issuecomment-1819032158

File: pkgs/development/compilers/gcc/patches/12/mangle-NIX_STORE-in-__FILE__.patch [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+Typical examples are `nix` -> `nlohmann_json` and `pipewire` ->
+`lttng-ust.dev`.
+
+Ideally we would like to use `-fmacro-prefix-map=` feature of `gcc` as:
+
+  -fmacro-prefix-map=/nix/store/$hash1-nlohmann-json-ver=/nix/store/eeee.eee-nlohmann-json-ver
+  -fmacro-prefix-map=/nix/...
+
+In practice it quickly exhausts argument lengtth limit due to `gcc`
```
Comment (dotlambda) [inline pkgs/development/compilers/gcc/patches/12/mangle-NIX_STORE-in-__FILE__.patch:20]:
```suggestion
In practice it quickly exhausts argument length limit due to `gcc`
```

File: pkgs/development/compilers/gcc/patches/12/mangle-NIX_STORE-in-__FILE__.patch [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+deficiency: https://gcc.gnu.org/PR111527
+
+Until it;s fixed let's hardcode header mangling if $NIX_STORE variable
```
Comment (dotlambda) [inline pkgs/development/compilers/gcc/patches/12/mangle-NIX_STORE-in-__FILE__.patch:23]:
```suggestion
Until it's fixed let's hardcode header mangling if $NIX_STORE variable
```

**Comment by trofi** [comment]
> This broke `python3.pkgs.deltachat`: [#267870 (comment)](https://github.com/NixOS/nixpkgs/pull/267870#issuecomment-1819032158)

Good find! Proposed the fix as https://github.com/NixOS/nixpkgs/pull/268920

File: pkgs/development/compilers/gcc/patches/12/mangle-NIX_STORE-in-__FILE__.patch (hunk 4 above)
Comment (trofi) [reply pkgs/development/compilers/gcc/patches/12/mangle-NIX_STORE-in-__FILE__.patch:20]: Good catch! Proposed the fix as https://github.com/NixOS/nixpkgs/pull/268923

File: pkgs/development/compilers/gcc/patches/12/mangle-NIX_STORE-in-__FILE__.patch (hunk 5 above)
Comment (trofi) [reply pkgs/development/compilers/gcc/patches/12/mangle-NIX_STORE-in-__FILE__.patch:23]: Good catch! Proposed the fix as https://github.com/NixOS/nixpkgs/pull/268923

**Pull request merged**